File: model/src/test/java/com/github/fburato/highwheelmodules/model/modules/TrackingModuleDependencyTest.java
package com.github.fburato.highwheelmodules.model.modules;

import org.junit.Test;
import com.github.fburato.highwheelmodules.model.bytecode.AccessPoint;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

import static org.assertj.core.api.Assertions.assertThat;

public class TrackingModuleDependencyTest {

  private final HWModule moduleA = HWModule.make("module A", "module A").get();
  private final HWModule moduleB = HWModule.make("module B", "module B").get();
  private final AccessPoint exampleSource = AccessPoint.create(ElementName.fromString("A"));
  private final AccessPoint exampleSource1 = AccessPoint.create(ElementName.fromString("A1"));
  private final AccessPoint exampleDest = AccessPoint.create(ElementName.fromString("B"));
  private final AccessPoint exampleDest1 = AccessPoint.create(ElementName.fromString("A1"));
  private final TrackingModuleDependency testee = new TrackingModuleDependency(moduleA, moduleB);

  @Test
  public void evidencesShouldBeEmptyOnEmptyTrackingDependency() {
    assertThat(testee.getSources().isEmpty()).isTrue();
    assertThat(testee.getDestinations().isEmpty()).isTrue();
  }

  @Test
  public void sourcesShouldReturnAddedEvidences() {
    testee.addEvidence(exampleSource, exampleDest);

    assertThat(testee.getSources()).containsExactly(exampleSource);
    assertThat(testee.getDestinations()).containsExactly(exampleDest);
  }

  @Test
  public void getDestinationShouldReturnEmptyOnNotAddedEvidences() {
    assertThat(testee.getDestinationsFromSource(exampleSource).isEmpty()).isTrue();
  }

  @Test
  public void getDestinationShouldReturnEmptyOnEvidenceOnDifferentSource() {
    testee.addEvidence(exampleDest, exampleDest);

    assertThat(testee.getDestinationsFromSource(exampleSource).isEmpty()).isTrue();
  }

  @Test
  public void getDestinationShouldReturnExpectedEvidences() {
    testee.addEvidence(exampleSource, exampleDest);

    assertThat(testee.getDestinationsFromSource(exampleSource)).containsExactly(exampleDest);
  }

  @Test
  public void shouldSaveAllEvidenceFromSameSource() {
    final TrackingModuleDependency testee = new TrackingModuleDependency(moduleA, moduleB);
    testee.addEvidence(exampleSource, exampleDest);
    testee.addEvidence(exampleSource, exampleDest1);

    assertThat(testee.getSources()).containsExactlyInAnyOrder(exampleSource);
    assertThat(testee.getDestinations()).containsExactlyInAnyOrder(exampleDest, exampleDest1);
    assertThat(testee.getEvidenceCounter()).isEqualTo(2);
  }

  @Test
  public void shouldSaveAllEvidenceDifferentSources() {
    final TrackingModuleDependency testee = new TrackingModuleDependency(moduleA, moduleB);
    testee.addEvidence(exampleSource, exampleDest);
    testee.addEvidence(exampleSource, exampleDest1);
    testee.addEvidence(exampleSource1, exampleDest);
    testee.addEvidence(exampleSource1, exampleDest1);

    assertThat(testee.getSources()).containsExactlyInAnyOrder(exampleSource, exampleSource1);
    assertThat(testee.getDestinations()).containsExactlyInAnyOrder(exampleDest, exampleDest1);
    assertThat(testee.getEvidenceCounter()).isEqualTo(4);
  }
}

File: model/src/test/scala/com/github/fburato/highwheelmodules/model/modules/TrackingModuleDependencyTest.scala
package com.github.fburato.highwheelmodules.model.modules

import com.github.fburato.highwheelmodules.model.bytecode.{AccessPoint, ElementName}
import org.apache.commons.lang3.RandomStringUtils
import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class TrackingModuleDependencyTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  private def randomString: String = RandomStringUtils.randomAlphanumeric(10)

  private val moduleA = HWModule.make(randomString, List(randomString)).get
  private val moduleB = HWModule.make(randomString, List(randomString)).get
  private val exampleSource1 = AccessPoint(ElementName.fromString(randomString))
  private val exampleSource2 = AccessPoint(ElementName.fromString(randomString))
  private val exampleDest1 = AccessPoint(ElementName.fromString(randomString))
  private val exampleDest2 = AccessPoint(ElementName.fromString(randomString))
  private val testee = TrackingModuleDependency(moduleA, moduleB)

  "evidences should be empty on empty tracking dependency" in {
    testee.sources shouldEqual Set()
    testee.destinations shouldEqual Set()
  }

  "sources should return added evidence" in {
    testee.addEvidence(exampleSource1, exampleDest1)

    testee.sources should contain theSameElementsAs Set(exampleSource1)
    testee.destinations should contain theSameElementsAs Set(exampleDest1)
  }

  "getDestinationFromSources" should {
    "be empty on no evidence present" in {
      testee.destinationsFromSource(exampleSource1) shouldEqual Set()
    }

    "be empty evidence if no evidence added with requested source" in {
      testee.addEvidence(exampleSource2, exampleDest1)

      testee.destinationsFromSource(exampleSource1) shouldEqual Set()
    }

    "return the expected evidence" in {
      testee.addEvidence(exampleSource1, exampleDest1)

      testee.destinationsFromSource(exampleSource1) should contain theSameElementsAs Set(
        exampleDest1
      )
    }
  }

  "addEvidence" should {
    val testee = TrackingModuleDependency(moduleA, moduleB)

    "save all evidence from same source" in {
      testee.addEvidence(exampleSource1, exampleDest1)
      testee.addEvidence(exampleSource1, exampleDest2)

      testee.sources should contain theSameElementsAs Set(exampleSource1)
      testee.destinations should contain theSameElementsAs Set(exampleDest1, exampleDest2)
      testee.evidenceCounter shouldEqual 2
    }

    "save all evidence from different sources" in {
      testee.addEvidence(exampleSource1, exampleDest1)
      testee.addEvidence(exampleSource1, exampleDest2)
      testee.addEvidence(exampleSource2, exampleDest1)
      testee.addEvidence(exampleSource2, exampleDest2)

      testee.sources should contain theSameElementsAs Set(exampleSource1, exampleSource2)
      testee.destinations should contain theSameElementsAs Set(exampleDest1, exampleDest2)
      testee.evidenceCounter shouldEqual 4
    }
  }
}
